File: apps/easytimepro/easytimepro/easy_timepro/report/easy_timepro_daily_attendance/easy_timepro_daily_attendance.py
# Copyright (c) 2026, Tejas and contributors
# MIT License

from __future__ import annotations

from collections import defaultdict

import frappe
from frappe import _
from frappe.utils import format_datetime, formatdate, get_datetime, getdate, time_diff_in_hours


def execute(filters: dict | None = None):
	filters = frappe._dict(filters or {})
	columns = get_columns()
	data = get_data(filters)
	chart = get_chart(data)
	summary = get_summary(data)
	message = get_message(filters, data)
	return columns, data, message, chart, summary


def get_message(filters: frappe._dict, data: list[dict]) -> str:
	from_date = formatdate(filters.from_date) if filters.from_date else ""
	to_date = formatdate(filters.to_date) if filters.to_date else ""
	return f"""
	<div class="etp-report-banner">
		<div class="etp-report-banner__title">{_("Easy TimePro Daily Attendance")}</div>
		<div class="etp-report-banner__meta">
			{_('Period')}: <b>{from_date}</b> → <b>{to_date}</b>
			&nbsp;·&nbsp; {_('Records')}: <b>{len(data)}</b>
		</div>
	</div>
	"""


def get_columns() -> list[dict]:
	return [
		{
			"label": _("Date"),
			"fieldname": "attendance_date",
			"fieldtype": "Date",
			"width": 120,
		},
		{
			"label": _("Employee Name"),
			"fieldname": "employee_name",
			"fieldtype": "Data",
			"width": 220,
		},
		{
			"label": _("Employee"),
			"fieldname": "employee",
			"fieldtype": "Link",
			"options": "Employee",
			"width": 200,
		},
		{
			"label": _("Employee ID"),
			"fieldname": "emp_code",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": _("Department"),
			"fieldname": "department",
			"fieldtype": "Data",
			"width": 200,
		},
		{
			"label": _("Check In"),
			"fieldname": "first_in",
			"fieldtype": "Data",
			"width": 130,
		},
		{
			"label": _("Check Out"),
			"fieldname": "last_out",
			"fieldtype": "Data",
			"width": 130,
		},
		{
			"label": _("Work Hours"),
			"fieldname": "work_hours",
			"fieldtype": "Float",
			"width": 150,
			"precision": 2,
		},
		{
			"label": _("Status"),
			"fieldname": "day_status",
			"fieldtype": "Data",
			"width": 150,
		},
	]


def get_data(filters: frappe._dict) -> list[dict]:
	from_date = getdate(filters.from_date) if filters.from_date else getdate()
	to_date = getdate(filters.to_date) if filters.to_date else getdate()

	conditions = ["punch_time >= %(from_dt)s", "punch_time < %(to_dt)s"]
	params = {
		"from_dt": f"{from_date} 00:00:00",
		"to_dt": f"{frappe.utils.add_days(to_date, 1)} 00:00:00",
	}

	if filters.employee:
		conditions.append("employee = %(employee)s")
		params["employee"] = filters.employee
	if filters.emp_code:
		conditions.append("emp_code = %(emp_code)s")
		params["emp_code"] = filters.emp_code

	where = " and ".join(conditions)
	rows = frappe.db.sql(
		f"""
		select
			name, emp_code, employee, employee_name, punch_time, log_type
		from `tabEasy TimePro Punch Log`
		where {where}
		order by punch_time asc
		""",
		params,
		as_dict=True,
	)

	grouped: dict[tuple, list] = defaultdict(list)
	for row in rows:
		day = getdate(row.punch_time)
		key = (day, row.employee or "", row.emp_code or "")
		grouped[key].append(row)

	employee_ids = {row.employee for row in rows if row.employee}
	employee_meta = {}
	if employee_ids:
		for emp in frappe.get_all(
			"Employee",
			filters={"name": ("in", list(employee_ids))},
			fields=["name", "employee_name", "department"],
		):
			employee_meta[emp.name] = emp

	data = []
	for (day, employee, emp_code), punches in sorted(
		grouped.items(), key=lambda x: (x[0][0], x[0][2]), reverse=True
	):
		ins = [p for p in punches if p.log_type == "IN"]
		outs = [p for p in punches if p.log_type == "OUT"]
		first_in = ins[0].punch_time if ins else None
		last_out = outs[-1].punch_time if outs else None

		work_hours = 0.0
		if first_in and last_out and get_datetime(last_out) > get_datetime(first_in):
			work_hours = max(0.0, time_diff_in_hours(last_out, first_in))

		if first_in and last_out:
			status = "Complete"
		elif first_in:
			status = "Checked In"
		elif last_out:
			status = "Checked Out"
		else:
			status = "No IN/OUT"

		meta = employee_meta.get(employee) or frappe._dict()
		employee_name = punches[0].employee_name or meta.employee_name or ""

		data.append(
			{
				"attendance_date": day,
				"emp_code": emp_code,
				"employee": employee or None,
				"employee_name": employee_name,
				"department": meta.department or "",
				"first_in": format_datetime(first_in, "HH:mm") if first_in else "—",
				"last_out": format_datetime(last_out, "HH:mm") if last_out else "—",
				"in_count": len(ins),
				"out_count": len(outs),
				"total_punches": len(punches),
				"work_hours": round(work_hours, 2),
				"day_status": status,
			}
		)

	return data


def get_chart(data: list[dict]) -> dict | None:
	if not data:
		return None

	by_date: dict = defaultdict(lambda: {"in": 0, "out": 0, "hours": 0.0})
	for row in data:
		key = formatdate(row["attendance_date"], "dd MMM")
		by_date[key]["in"] += row["in_count"]
		by_date[key]["out"] += row["out_count"]
		by_date[key]["hours"] += row["work_hours"]

	# preserve chronological order from sorted dates
	ordered = []
	seen = set()
	for row in sorted(data, key=lambda r: r["attendance_date"]):
		label = formatdate(row["attendance_date"], "dd MMM")
		if label not in seen:
			seen.add(label)
			ordered.append(label)
	labels = ordered[-14:]

	return {
		"data": {
			"labels": labels,
			"datasets": [
				{"name": _("Check In"), "values": [by_date[d]["in"] for d in labels]},
				{"name": _("Check Out"), "values": [by_date[d]["out"] for d in labels]},
			],
		},
		"type": "line",
		"lineOptions": {"regionFill": 1, "hideDots": 0, "heatline": 0},
		"axisOptions": {"xIsSeries": 1},
		"colors": ["#1f4b99", "#b42318"],
		"height": 280,
	}


def get_summary(data: list[dict]) -> list[dict]:
	employees = {r["employee"] or r["emp_code"] for r in data}
	complete = sum(1 for r in data if r["day_status"] == "Complete")
	checked_in = sum(1 for r in data if r["day_status"] == "Checked In")
	total_punches = sum(r["total_punches"] for r in data)
	avg_hours = (
		round(sum(r["work_hours"] for r in data if r["work_hours"]) / max(1, complete), 2)
		if complete
		else 0
	)

	return [
		{"value": len(data), "label": _("Attendance Days"), "datatype": "Int"},
		{"value": len(employees), "label": _("Employees"), "datatype": "Int"},
		{"value": complete, "label": _("Complete Days"), "datatype": "Int", "indicator": "Green"},
		{"value": checked_in, "label": _("Still Checked In"), "datatype": "Int", "indicator": "Blue"},
		{"value": total_punches, "label": _("Total Punches"), "datatype": "Int"},
		{"value": avg_hours, "label": _("Avg Work Hours"), "datatype": "Float", "indicator": "Orange"},
	]

File: apps/easytimepro/easytimepro/easy_timepro/report/easy_timepro_daily_attendance/easy_timepro_daily_attendance.js
// Copyright (c) 2026, Tejas and contributors
// MIT License

frappe.query_reports["Easy TimePro Daily Attendance"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_start(),
			reqd: 1,
			width: "80px",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
			width: "80px",
		},
		{
			fieldname: "employee",
			label: __("Employee"),
			fieldtype: "Link",
			options: "Employee",
			get_query: () => ({
				filters: { status: "Active" },
			}),
		},
		{
			fieldname: "emp_code",
			label: __("Employee ID"),
			fieldtype: "Data",
		},
	],

	get_datatable_options(options) {
		options.layout = "fluid";
		options.cellHeight = 36;
		options.columns = (options.columns || []).map((col) => {
			const id = col.id || col.fieldname;
			if (id === "employee") {
				return Object.assign({}, col, { width: 200 });
			}
			if (id === "emp_code") {
				// Keep Employee ID compact; don't stretch with fluid layout
				return Object.assign({}, col, { width: 100, resizable: false });
			}
			return col;
		});
		return options;
	},

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (!data) {
			return value;
		}

		if (column.fieldname === "employee_name") {
			return `<span class="etp-name">${frappe.utils.escape_html(data.employee_name || "")}</span>`;
		}

		if (column.fieldname === "emp_code") {
			return `<span class="etp-code">${frappe.utils.escape_html(data.emp_code || "")}</span>`;
		}

		if (column.fieldname === "department") {
			if (!data.department) {
				return `<span class="etp-muted">—</span>`;
			}
			return `<span class="etp-dept">${frappe.utils.escape_html(data.department)}</span>`;
		}

		if (column.fieldname === "day_status") {
			const tone =
				{
					Complete: "green",
					"Checked In": "blue",
					"Checked Out": "orange",
					"No IN/OUT": "gray",
				}[data.day_status] || "gray";
			return `<span class="etp-pill etp-pill--${tone}">${frappe.utils.escape_html(
				data.day_status || ""
			)}</span>`;
		}

		if (column.fieldname === "first_in") {
			if (!data.first_in || data.first_in === "—") {
				return `<span class="etp-muted">—</span>`;
			}
			return `<span class="etp-punch etp-punch--in"><span class="etp-punch__label">IN</span>${frappe.utils.escape_html(
				data.first_in
			)}</span>`;
		}

		if (column.fieldname === "last_out") {
			if (!data.last_out || data.last_out === "—") {
				return `<span class="etp-muted">—</span>`;
			}
			return `<span class="etp-punch etp-punch--out"><span class="etp-punch__label">OUT</span>${frappe.utils.escape_html(
				data.last_out
			)}</span>`;
		}

		if (column.fieldname === "work_hours") {
			const hours = Number(data.work_hours || 0);
			if (!hours) {
				return `<span class="etp-muted">0.00h</span>`;
			}
			const tone = hours >= 8 ? "good" : hours >= 4 ? "ok" : "low";
			const width = Math.min(100, Math.round((hours / 10) * 100));
			return `<div class="etp-hours">
				<div class="etp-hours__bar"><span style="width:${width}%" class="etp-hours__fill etp-hours__fill--${tone}"></span></div>
				<span class="etp-hours__val">${hours.toFixed(2)}h</span>
			</div>`;
		}

		return value;
	},

	onload(report) {
		const old = document.getElementById("etp-daily-att-style");
		if (old) {
			old.remove();
		}
		if (!document.getElementById("etp-daily-att-style-v2")) {
			const style = document.createElement("style");
			style.id = "etp-daily-att-style-v2";
			style.textContent = `
				.etp-report-banner {
					margin: 4px 0 14px;
					padding: 14px 16px;
					border-radius: 10px;
					border: 1px solid #e4e7ec;
					border-left: 4px solid #1f4b99;
					background: linear-gradient(180deg, #ffffff 0%, #f8fafc 100%);
					box-shadow: 0 1px 2px rgba(16,24,40,0.04);
				}
				.etp-report-banner__title {
					font-size: 15px;
					font-weight: 700;
					color: #101828;
					margin-bottom: 4px;
				}
				.etp-report-banner__meta {
					font-size: 12px;
					color: #667085;
				}
				.etp-name { font-weight: 650; color: #101828; }
				.etp-code {
					font-variant-numeric: tabular-nums;
					font-size: 12px;
					color: #475467;
					background: #f2f4f7;
					border: 1px solid #e4e7ec;
					border-radius: 6px;
					padding: 1px 7px;
				}
				.etp-dept { color: #344054; font-size: 12.5px; }
				.etp-muted { color: #98a2b3; }
				.etp-pill {
					display: inline-flex;
					align-items: center;
					padding: 1px 8px;
					border-radius: 999px;
					font-size: 11px;
					font-weight: 600;
					line-height: 1.35;
					border: 1px solid transparent;
					white-space: nowrap;
					vertical-align: middle;
				}
				.etp-pill--green { background: #ecfdf3; color: #067647; border-color: #abefc6; }
				.etp-pill--blue { background: #eff4ff; color: #175cd3; border-color: #b2ccff; }
				.etp-pill--orange { background: #fffaeb; color: #b54708; border-color: #fedf89; }
				.etp-pill--gray { background: #f2f4f7; color: #344054; border-color: #d0d5dd; }
				.etp-punch {
					display: inline-flex;
					align-items: center;
					gap: 6px;
					font-weight: 650;
					font-variant-numeric: tabular-nums;
					font-size: 13px;
				}
				.etp-punch__label {
					font-size: 10px;
					font-weight: 700;
					letter-spacing: 0.04em;
					padding: 1px 5px;
					border-radius: 4px;
				}
				.etp-punch--in { color: #175cd3; }
				.etp-punch--in .etp-punch__label { background: #eff4ff; color: #175cd3; }
				.etp-punch--out { color: #b42318; }
				.etp-punch--out .etp-punch__label { background: #fef3f2; color: #b42318; }
				.etp-hours {
					display: flex;
					align-items: center;
					gap: 8px;
					width: 100%;
					min-width: 110px;
					max-width: 180px;
				}
				.etp-hours__bar {
					flex: 1;
					height: 6px;
					border-radius: 99px;
					background: #eef2f6;
					overflow: hidden;
				}
				.etp-hours__fill {
					display: block;
					height: 100%;
					border-radius: 99px;
				}
				.etp-hours__fill--good { background: #12b76a; }
				.etp-hours__fill--ok { background: #2e90fa; }
				.etp-hours__fill--low { background: #f79009; }
				.etp-hours__val {
					font-weight: 700;
					font-variant-numeric: tabular-nums;
					color: #1f4b99;
					font-size: 12px;
					min-width: 42px;
				}
				.report-wrapper .datatable,
				.report-wrapper .dt-scrollable {
					width: 100% !important;
				}
				.report-summary {
					gap: 10px !important;
				}
				.report-summary > div {
					border-radius: 10px !important;
					border: 1px solid #e4e7ec !important;
					box-shadow: 0 1px 2px rgba(16,24,40,0.04);
					background: #fff !important;
				}
			`;
			document.head.appendChild(style);
		}

		report.page.add_inner_button(__("Punch Logs"), () => {
			frappe.set_route("List", "Easy TimePro Punch Log");
		});
		report.page.add_inner_button(__("Employee Checkin"), () => {
			frappe.set_route("List", "Employee Checkin");
		});
	},
};
